### ISSUE TITLE
Move packages to openwrt/openwrt to fix warnings?

### ISSUE THREAD

**Issue by PolynomialDivision** [openwrt/openwrt#10532]
Currently OpenWrt experiences warnings:
```
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a build dependency on 'libpam', which does not exist
WARNING: Makefile 'package/boot/kexec-tools/Makefile' has a dependency on 'liblzma', which does not exist
WARNING: Makefile 'package/network/services/lldpd/Makefile' has a dependency on 'libnetsnmp', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a build dependency on 'libpam', which does not exist
```

- busybox and policycoreutils depend on libpam resulting in warnings (cc @stintel)
- kexec-tools depend liblzma (no Maintainer)
- lldpd depend on libnetsnmp (cc @stintel)

**Comment by ynezz** [comment]
> Move packages to openwrt/openwrt to fix warnings?

IMO this is not a proper fix. What about fixing `scripts/package-metadata.pl` (or some other relevant parts) instead? That warning message is little bit misleading, there is no such dependency unless you've that specific config knob enabled.

**Comment by PolynomialDivision** [comment]
> That warning message is little bit misleading, there is no such dependency unless you've that specific config knob enabled.

But if you enable it, you have the dependency from openwrt/openwrt to openwrt/packages. I thought openwrt main repository would be structured like a hierarchical tree that has no dependencies to its leafs.

**Comment by ynezz** [comment]
> I thought openwrt main repository would be structured like a hierarchical tree that has no dependencies to its leafs.

Indeed, and AFAIK it works like that, you don't need any external dependency for default and core configuration. It builds and its usable without any other feeds.

> But if you enable it, you have the dependency from openwrt/openwrt to openwrt/packages.

IMO this is acceptable given current build system architecture and layout.

**Comment by PolynomialDivision** [comment]
Okay, than lets fix the scripts. :)

**Comment by anand16158** [comment]
Root Cause
The warnings come from scripts/package-metadata.pl in the gen_package_mk() function. There are two places that emit warnings:

1. Runtime dependencies (around line 464):

unless (defined $vpkg_dep) {
    warn sprintf "WARNING: Makefile '%s' has a dependency on '%s', which does not exist\n",
        $src->{makefile}, $dep;
    next;
}
2. Build dependencies (around line 541):

unless (defined($src_dep) && (!$deptype || grep { $_ eq $deptype } @{$src_dep->{buildtypes}})) {
    warn sprintf "WARNING: Makefile '%s' has a build dependency on '%s', which does not exist\n",
        $src->{makefile}, $dep.$depsuffix;
    next;
}
The problem is that all the affected dependencies are conditional (gated behind a config option):

Makefile	Dependency	Condition
busybox	libpam	BUSYBOX_CONFIG_PAM
kexec-tools	liblzma	KEXEC_LZMA
lldpd	libnetsnmp	LLDPD_WITH_SNMP
policycoreutils	libpam	BUSYBOX_CONFIG_PAM
These packages (libpam, liblzma, libnetsnmp) come from the external packages feed, not the core repo. The script already parses the condition into $condition but still warns even when the dependency is conditional and the feed isn't installed.

The Fix
File: scripts/package-metadata.pl

Add next if $condition; before each warning so conditional dependencies from missing feeds are silently skipped.
